### PR TITLE
haskellPackages: ghc 8.10.4 -> 8.10.6

### PR DESCRIPTION
**Pull request by maralorn**
We are doing a set rebuild with ghc 8.10.6. I have not done a hackage bump so that we can easily see breakage.

---

### This Merge

This PR is the regular merge of the `haskell-updates` branch into `master`.

This branch is being continually built and tested by hydra at https://hydra.nixos.org/jobset/nixpkgs/haskell-updates.

I will aim to merge this PR **by 2021-08-22**. If I can merge it earlier, there might be successor PRs in that time window. As part of our rotation @expipiplus1 will continue these merges from 2021-08-23 to 2021-09-06.

### haskellPackages Workflow Summary

Our workflow is currently described in
[`pkgs/development/haskell-modules/HACKING.md`](https://github.com/NixOS/nixpkgs/blob/haskell-updates/pkgs/development/haskell-modules/HACKING.md).

The short version is this:
* We regularly update the Stackage and Hackage pins on `haskell-updates` (normally at the beginning of a merge window).
* The community fixes builds of Haskell packages on that branch.
* We aim at at least one merge of `haskell-updates` into `master` every two weeks.
* We only do the merge if the [`mergeable`](https://hydra.nixos.org/job/nixpkgs/haskell-updates/mergeable) job is succeeding on hydra.
* If a [`maintained`](https://hydra.nixos.org/job/nixpkgs/haskell-updates/maintained) package is still broken at the time of merge, we will only merge if the maintainer has been pinged 7 days in advance. (If you care about a Haskell package, become a maintainer!)

---

This is the follow-up to #133422.

### CODE REVIEW

**Comment by keke-cute** [comment]
Will support aarch64-darwin?

**Comment by sternenseemann** [comment]
> Will support aarch64-darwin?

For cross-compilation yes (from `x86_64-darwin`, at least in theory), natively *no*. We don't have a `ghc8106Binary` derivation at the moment, so we can't bootstrap GHC 8.10.6 on aarch64-darwin natively.

None of the Haskell maintainers has any aarch64-darwin hardware for testing purposes, so if you want it, please contribute it (or, I guess, mail us an M1 macbook).

**Comment by dbaynard** [comment]
> None of the Haskell maintainers has any aarch64-darwin hardware for testing purposes

@sternenseemann What do you need, for testing?

**Comment by maralorn** [comment]
There was one significant breakage in ghc-api-compat everything else seems fine. Since that breakage can hopefully be fixed by bumping hackage I will proceed to do so.

**Comment by cdepillabout** [comment]
When building haskell packages (I noticed it when building `gitit`), I'm seeing messages like these in the logs:

```
/nix/store/hy3lz2vfv9qq2v5jz9nzlx6mmiaq79rj-binutils-2.35.1/bin/ld.gold: warning: /nix/store/cvr0kjg2q7z2wwhjblx6c73rv422k8cm-glibc-2.33-47/lib/crt1.o: unknown program property type 0xc00080
02 in .note.gnu.property section
```

It doesn't appear to affect the build or the final binary that is output.

Is this something that happened from the 8.10.6 update?

**Comment by sternenseemann** [comment]
No has been happening for a bit, not sure when exactly it started (I guess we would need to do a bisect). I've been meaning to open an issue asking around how harmless this is, but have been too lazy so far.

**Comment by sternenseemann** [comment]
> No has been happening for a bit, not sure when exactly it started (I guess we would need to do a bisect). I've been meaning to open an issue asking around how harmless this is, but have been too lazy so far.

Current suspicion is acdcb8566ae560752a8a1ce3aacc617a3da6cdd5, but it's hard to confirm unfortunately.

**Comment by sternenseemann** [comment]
* Upstream bug in binutils <https://sourceware.org/bugzilla/show_bug.cgi?id=22914>
* Upstream fix <https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git;h=6c04fd9b2fb4396c0189cb414ce598161ac8673e>

**Comment by prusnak** [comment]
> None of the Haskell maintainers has any aarch64-darwin hardware for testing purposes, so if you want it, please contribute it (or, I guess, mail us an M1 macbook).

I submitted my aarch64-darwin work in PR https://github.com/NixOS/nixpkgs/pull/135345
